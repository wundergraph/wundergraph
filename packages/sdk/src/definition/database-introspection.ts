import { IntrospectionCacheConfiguration, introspectWithCache } from './introspection-cache';
import { DatabaseSchema, mongodb, mysql, planetscale, postgresql, prisma, sqlite, sqlserver } from '../db/types';
import {
	Api,
	ApiIntrospectionOptions,
	ApiType,
	DatabaseApiCustom,
	DatabaseIntrospection,
	DataSource,
	MongoDBApi,
	MySQLApi,
	PlanetscaleApi,
	PostgresqlApi,
	PrismaApi,
	PrismaIntrospection,
	SQLiteApi,
	SQLServerApi,
} from './index';
import { introspectPrismaDatabaseWithRetries } from '../db/introspection';
import { buildSchema, parse, print } from 'graphql/index';
import { configuration } from '../graphql/configuration';
import {
	applyNameSpaceToFieldConfigurations,
	applyNameSpaceToGraphQLSchema,
	applyNameSpaceToTypeFields,
	generateTypeConfigurationsForNamespace,
} from './namespacing';
import { mapInputVariable, resolveVariable } from '../configure/variables';
import { DataSourceKind, FieldConfiguration, SingleTypeField, TypeConfiguration } from '@wundergraph/protobuf';
import { fileHash, urlHash } from '../localcache';

const introspectDatabase = async (
	introspection: DatabaseIntrospection,
	databaseSchema: DatabaseSchema,
	maxRetries: number
) => {
	const {
		success,
		message,
		graphql_schema,
		prisma_schema,
		interpolateVariableDefinitionAsJSON,
		jsonTypeFields,
		jsonResponseFields,
	} = await introspectPrismaDatabaseWithRetries(introspection, databaseSchema, maxRetries);
	if (!success) {
		return Promise.reject(message);
	}
	const schemaDocumentNode = parse(graphql_schema);
	const schema = print(schemaDocumentNode);
	const { RootNodes, ChildNodes, Fields } = configuration(schemaDocumentNode, {
		url: '',
		schemaExtension: introspection.schemaExtension,
	});
	const jsonFields = [...jsonTypeFields, ...jsonResponseFields];
	jsonFields.forEach((field) => {
		const fieldConfig = Fields.find((f) => f.typeName == field.typeName && f.fieldName == field.fieldName);
		if (fieldConfig) {
			fieldConfig.unescapeResponseJson = true;
		} else {
			Fields.push({
				fieldName: field.fieldName,
				typeName: field.typeName,
				unescapeResponseJson: true,
				argumentsConfiguration: [],
				path: [],
				requiresFields: [],
				disableDefaultFieldMapping: false,
			});
		}
	});
	const graphQLSchema = buildSchema(schema);
	const dataSource: DataSource<DatabaseApiCustom> = {
		Kind: databaseSchemaToKind(databaseSchema),
		RootNodes: applyNameSpaceToTypeFields(RootNodes, graphQLSchema, introspection.apiNamespace),
		ChildNodes: applyNameSpaceToTypeFields(ChildNodes, graphQLSchema, introspection.apiNamespace),
		Custom: {
			prisma_schema: prisma_schema,
			databaseURL: mapInputVariable(introspection.databaseURL),
			graphql_schema: schema,
			jsonTypeFields: applyNameSpaceToSingleTypeFields(jsonTypeFields, introspection.apiNamespace),
			jsonInputVariables: applyNameSpaceToTypeNames(interpolateVariableDefinitionAsJSON, introspection.apiNamespace),
		},
		Directives: [],
		RequestTimeoutSeconds: introspection.requestTimeoutSeconds ?? 0,
	};
	const dataSources: DataSource<DatabaseApiCustom>[] = [];
	dataSource.RootNodes.forEach((rootNode) => {
		rootNode.fieldNames.forEach((field) => {
			dataSources.push({
				...Object.assign({}, dataSource),
				RootNodes: [
					{
						typeName: rootNode.typeName,
						fieldNames: [field],
					},
				],
			});
		});
	});
	return {
		schema: applyNameSpaceToGraphQLSchema(schema, [], introspection.apiNamespace),
		dataSources: dataSources,
		fields: applyNameSpaceToFieldConfigurations(Fields, graphQLSchema, [], introspection.apiNamespace),
		types: generateTypeConfigurationsForNamespace(schema, introspection.apiNamespace),
		interpolateVariableDefinitionAsJSON: applyNameSpaceToTypeNames(
			interpolateVariableDefinitionAsJSON,
			introspection.apiNamespace
		),
	};
};

const databaseSchemaToKind = (schema: DatabaseSchema): DataSourceKind => {
	switch (schema) {
		case planetscale:
			return DataSourceKind.MYSQL;
		case mysql:
			return DataSourceKind.MYSQL;
		case postgresql:
			return DataSourceKind.POSTGRESQL;
		case sqlite:
			return DataSourceKind.SQLITE;
		case sqlserver:
			return DataSourceKind.SQLSERVER;
		case mongodb:
			return DataSourceKind.MONGODB;
		case prisma:
			return DataSourceKind.PRISMA;
		default:
			throw new Error(`databaseSchemaToKind not implemented for: ${schema}`);
	}
};

const applyNameSpaceToSingleTypeFields = (typeFields: SingleTypeField[], namespace?: string): SingleTypeField[] => {
	if (!namespace) {
		return typeFields;
	}
	return typeFields.map((typeField) => ({
		...typeField,
		typeName: `${namespace}_${typeField.typeName}`,
	}));
};

const applyNameSpaceToTypeNames = (typeNames: string[], namespace?: string): string[] => {
	if (!namespace) {
		return typeNames;
	}
	return typeNames.map((typeName) => {
		return `${namespace}_${typeName}`;
	});
};

type databaseConstructor<T> = (
	schema: string,
	namespace: string,
	dataSources: DataSource<DatabaseApiCustom>[],
	fields: FieldConfiguration[],
	types: TypeConfiguration[],
	interpolateVariableDefinitionAsJSON: string[],
	customJsonScalars?: string[] | undefined
) => T;

type DbApi = Api<DatabaseApiCustom>;

function introspectDatabaseWithCache<TApi extends Api<ApiType>>(
	ctor: databaseConstructor<TApi>,
	databaseUrlSchema: DatabaseSchema
) {
	const generator = async (introspection: DatabaseIntrospection, _: ApiIntrospectionOptions) => {
		const { schema, fields, types, dataSources, interpolateVariableDefinitionAsJSON } = await introspectDatabase(
			introspection,
			databaseUrlSchema,
			5
		);
		return ctor(
			schema,
			introspection.apiNamespace || '',
			dataSources,
			fields,
			types,
			interpolateVariableDefinitionAsJSON
		);
	};

	return async (introspection: DatabaseIntrospection) => {
		const url = resolveVariable(introspection.databaseURL);
		const hash = await urlHash(url);
		// Introspecting a database might be expensive, even if it's a local network URL.
		// Consider all sources as remote to cache them more aggressively
		const cacheConfig = { keyInput: hash };
		return introspectWithCache(introspection, cacheConfig, generator);
	};
}

export const introspectPostgresql = introspectDatabaseWithCache(
	(
		schema: string,
		namespace: string,
		dataSources: DataSource<DatabaseApiCustom>[],
		fields: FieldConfiguration[],
		types: TypeConfiguration[],
		interpolateVariableDefinitionAsJSON: string[],
		customJsonScalars?: string[] | undefined
	) =>
		new PostgresqlApi(
			schema,
			namespace,
			dataSources,
			fields,
			types,
			interpolateVariableDefinitionAsJSON,
			customJsonScalars
		),
	postgresql
);
export const introspectMySQL = introspectDatabaseWithCache(
	(
		schema: string,
		namespace: string,
		dataSources: DataSource<DatabaseApiCustom>[],
		fields: FieldConfiguration[],
		types: TypeConfiguration[],
		interpolateVariableDefinitionAsJSON: string[],
		customJsonScalars?: string[] | undefined
	) =>
		new MySQLApi(schema, namespace, dataSources, fields, types, interpolateVariableDefinitionAsJSON, customJsonScalars),
	mysql
);
export const introspectPlanetScale = introspectDatabaseWithCache(
	(
		schema: string,
		namespace: string,
		dataSources: DataSource<DatabaseApiCustom>[],
		fields: FieldConfiguration[],
		types: TypeConfiguration[],
		interpolateVariableDefinitionAsJSON: string[],
		customJsonScalars?: string[] | undefined
	) =>
		new PlanetscaleApi(
			schema,
			namespace,
			dataSources,
			fields,
			types,
			interpolateVariableDefinitionAsJSON,
			customJsonScalars
		),
	planetscale
);
export const introspectSQLite = introspectDatabaseWithCache(
	(
		schema: string,
		namespace: string,
		dataSources: DataSource<DatabaseApiCustom>[],
		fields: FieldConfiguration[],
		types: TypeConfiguration[],
		interpolateVariableDefinitionAsJSON: string[],
		customJsonScalars?: string[] | undefined
	) =>
		new SQLiteApi(
			schema,
			namespace,
			dataSources,
			fields,
			types,
			interpolateVariableDefinitionAsJSON,
			customJsonScalars
		),
	sqlite
);
export const introspectSQLServer = introspectDatabaseWithCache(
	(
		schema: string,
		namespace: string,
		dataSources: DataSource<DatabaseApiCustom>[],
		fields: FieldConfiguration[],
		types: TypeConfiguration[],
		interpolateVariableDefinitionAsJSON: string[],
		customJsonScalars?: string[] | undefined
	) =>
		new SQLServerApi(
			schema,
			namespace,
			dataSources,
			fields,
			types,
			interpolateVariableDefinitionAsJSON,
			customJsonScalars
		),
	sqlserver
);
export const introspectMongoDB = introspectDatabaseWithCache(
	(
		schema: string,
		namespace: string,
		dataSources: DataSource<DatabaseApiCustom>[],
		fields: FieldConfiguration[],
		types: TypeConfiguration[],
		interpolateVariableDefinitionAsJSON: string[],
		customJsonScalars?: string[] | undefined
	) =>
		new MongoDBApi(
			schema,
			namespace,
			dataSources,
			fields,
			types,
			interpolateVariableDefinitionAsJSON,
			customJsonScalars
		),
	mongodb
);

export const introspectPrisma = async (introspection: PrismaIntrospection) => {
	const cacheConfig: IntrospectionCacheConfiguration = {
		keyInput: await fileHash(introspection.prismaFilePath),
		dataSource: 'localFilesystem',
	};
	return introspectWithCache(
		introspection,
		cacheConfig,
		async (introspection: PrismaIntrospection, _: ApiIntrospectionOptions): Promise<PrismaApi> => {
			const { schema, fields, types, dataSources, interpolateVariableDefinitionAsJSON } = await introspectDatabase(
				{ ...introspection, databaseURL: introspection.prismaFilePath },
				prisma,
				5
			);
			return new PrismaApi(
				schema,
				introspection.apiNamespace || '',
				dataSources,
				fields,
				types,
				interpolateVariableDefinitionAsJSON
			);
		}
	);
};
