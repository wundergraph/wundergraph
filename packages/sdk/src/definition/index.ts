import { buildClientSchema, GraphQLSchema, introspectionFromSchema, parse, print, printSchema } from 'graphql';
import { renameTypeFields, renameTypes } from '../graphql/renametypes';
import {
	ArgumentSource,
	ConfigurationVariable,
	DataSourceKind,
	DirectiveConfiguration,
	FetchConfiguration,
	FieldConfiguration,
	GraphQLDataSourceHooksConfiguration,
	MTLSConfiguration,
	SigningMethod,
	SingleTypeField,
	StatusCodeTypeMapping,
	TypeConfiguration,
	TypeField,
	UpstreamAuthentication,
	UpstreamAuthenticationKind,
} from '@wundergraph/protobuf';
import { applyNameSpaceToGraphQLSchema } from './namespacing';
import { InputVariable, mapInputVariable } from '../configure/variables';
import { introspectGraphqlWithCache } from './graphql-introspection';
import { introspectFederation } from './federation-introspection';
import { IGraphqlIntrospectionHeadersBuilder, IHeadersBuilder } from './headers-builder';
import { introspectOpenApi, introspectOpenApiV2 } from './openapi-introspection';
import {
	introspectMongoDB,
	introspectMySQL,
	introspectPlanetScale,
	introspectPostgresql,
	introspectPrisma,
	introspectSQLite,
	introspectSQLServer,
} from './database-introspection';
import { introspectSoap } from './soap-introspection';

// Use UPPERCASE for environment variables
export const WG_DATA_SOURCE_POLLING_MODE = process.env['WG_DATA_SOURCE_POLLING_MODE'] === 'true';
export const WG_ENABLE_INTROSPECTION_CACHE = process.env['WG_ENABLE_INTROSPECTION_CACHE'] === 'true';
export const WG_INTROSPECTION_CACHE_SKIP = process.env['WG_INTROSPECTION_CACHE_SKIP'] === 'true';
// Only use the introspection cache, return an error when hitting the network
export const WG_ENABLE_INTROSPECTION_OFFLINE = process.env['WG_ENABLE_INTROSPECTION_OFFLINE'] === 'true';

export const WG_PRETTY_GRAPHQL_VALIDATION_ERRORS = process.env['WG_PRETTY_GRAPHQL_VALIDATION_ERRORS'] === 'true';
// Default polling interval for data sources without one
export const WG_DATA_SOURCE_DEFAULT_POLLING_INTERVAL_SECONDS = (() => {
	const seconds = parseInt(process.env['WG_DATA_SOURCE_DEFAULT_POLLING_INTERVAL_SECONDS'] ?? '', 10);
	return isNaN(seconds) ? 0 : seconds;
})();

/**
 * ApiIntrospectionOptions contains options that are passed to
 * all ApiIntrospector<T> functions
 */
export interface ApiIntrospectionOptions {
	/**
	 * Global proxy URL, which might be overridden at the data source level
	 */
	httpProxyUrl?: string;
	apiID?: string;
}

export interface RenameType {
	from: string;
	to: string;
}

export interface RenameTypes {
	renameTypes: (rename: RenameType[]) => void;
}

export interface RenameTypeField {
	typeName: string;
	fromFieldName: string;
	toFieldName: string;
}

export interface RenameTypeFields {
	renameTypeFields: (rename: RenameTypeField[]) => void;
}

export type ApiType = GraphQLApiCustom | RESTApiCustom | DatabaseApiCustom;

export class Api<T = ApiType> implements RenameTypes, RenameTypeFields {
	constructor(
		schema: string,
		namespace: string,
		dataSources: DataSource<T>[],
		fields: FieldConfiguration[],
		types: TypeConfiguration[],
		interpolateVariableDefinitionAsJSON: string[],
		customJsonScalars?: string[]
	) {
		this.Schema = schema;
		this.Namespace = namespace;
		this.DataSources = dataSources;
		this.Fields = fields;
		this.Types = types;
		this.interpolateVariableDefinitionAsJSON = interpolateVariableDefinitionAsJSON;
		this.CustomJsonScalars = customJsonScalars;
	}

	DefaultFlushInterval: number = 500;
	Schema: string;
	DataSources: DataSource<T>[];
	Fields: FieldConfiguration[];
	Types: TypeConfiguration[];
	interpolateVariableDefinitionAsJSON: string[];
	CustomJsonScalars?: string[];
	Namespace: string;

	renameTypes(rename: RenameType[]): void {
		this.Schema = renameTypes(this.Schema, rename);
		this.DataSources = this.DataSources.map((d) => {
			return {
				...d,
				RootNodes: typeFieldsRenameType(d.RootNodes, rename),
				ChildNodes: typeFieldsRenameType(d.ChildNodes, rename),
			};
		});
		this.Fields = this.Fields.map((field) => {
			const re = rename.find((r) => r.from === field.typeName);
			return {
				...field,
				typeName: re !== undefined ? re.to : field.typeName,
			};
		});
	}

	renameTypeFields(rename: RenameTypeField[]): void {
		this.Schema = renameTypeFields(this.Schema, rename);
		this.DataSources = this.DataSources.map((d) => {
			return {
				...d,
				RootNodes: typeFieldsRenameTypeField(d.RootNodes, rename),
				ChildNodes: typeFieldsRenameTypeField(d.ChildNodes, rename),
			};
		});
		this.Fields = this.Fields.map((field) => {
			const re = rename.find((re) => re.typeName === field.typeName && re.fromFieldName === field.fieldName);
			if (re !== undefined) {
				return {
					...field,
					fieldName: re.toFieldName,
					path: field.path.map((item) => (item === field.fieldName ? re.toFieldName : item)),
				};
			}
			const sameTypeRenameFields = rename.filter((re) => re.typeName === field.typeName);
			return {
				...field,
				requiresFields: field.requiresFields.map((f) => {
					const re = sameTypeRenameFields.find((sameTypeField) => sameTypeField.fromFieldName === f);
					if (re !== undefined) {
						return re.toFieldName;
					}
					return f;
				}),
				argumentsConfiguration: field.argumentsConfiguration.map((arg) => {
					if (arg.sourceType === ArgumentSource.OBJECT_FIELD) {
						return {
							...arg,
							sourcePath: arg.sourcePath.map((item) => {
								const re = sameTypeRenameFields.find((sameTypeField) => sameTypeField.fromFieldName === item);
								if (re !== undefined) {
									return re.toFieldName;
								}
								return item;
							}),
						};
					}
					return arg;
				}),
			};
		});
	}
}

const typeFieldsRenameType = (fields: TypeField[], rename: RenameType[]): TypeField[] => {
	return fields.map((node) => {
		const re = rename.find((r) => r.from === node.typeName);
		return {
			...node,
			typeName: re !== undefined ? re.to : node.typeName,
		};
	});
};

const typeFieldsRenameTypeField = (fields: TypeField[], rename: RenameTypeField[]): TypeField[] => {
	return fields.map((node) => {
		return {
			...node,
			fieldNames: node.fieldNames.map((field) => {
				const re = rename.find((re) => re.typeName === node.typeName && re.fromFieldName === field);
				if (re !== undefined) {
					return re.toFieldName;
				}
				return field;
			}),
		};
	});
};

export const createMockApi = async (sdl: string, apiNamespace?: string): Promise<Api<any>> => {
	const schema = print(parse(sdl));
	return new GraphQLApi(applyNameSpaceToGraphQLSchema(schema, [], apiNamespace), apiNamespace || '', [], [], [], []);
};

export class GraphQLApi extends Api<GraphQLApiCustom> {}

export class RESTApi extends Api<RESTApiCustom> {}

export class PostgresqlApi extends Api<DatabaseApiCustom> {}

export class MySQLApi extends Api<DatabaseApiCustom> {}

export class PlanetscaleApi extends Api<DatabaseApiCustom> {}

export class SQLiteApi extends Api<DatabaseApiCustom> {}

export class SQLServerApi extends Api<DatabaseApiCustom> {}

export class MongoDBApi extends Api<DatabaseApiCustom> {}

export class PrismaApi extends Api<DatabaseApiCustom> {}

export interface DataSource<Custom = unknown> {
	Id?: string;
	Kind: DataSourceKind;
	RootNodes: TypeField[];
	ChildNodes: TypeField[];
	Custom: Custom;
	Directives: DirectiveConfiguration[];
	RequestTimeoutSeconds: number;
}

interface GraphQLIntrospectionOptions {
	// loadSchemaFromString allows you to skip the introspection process and load the GraphQL Schema from a string instead
	// this way, you can import a GraphQL Schema file or load the Schema in more flexible ways than relying on sending a GraphQL Introspection Query
	loadSchemaFromString?: string | (() => string);
	customFloatScalars?: string[];
	customIntScalars?: string[];
	customJSONScalars?: string[];
	// switching internal to true will mark the origin as an internal GraphQL API
	// this will forward the original request and user info to the internal upstream
	// so that the request context can be enriched
	internal?: boolean;
	skipRenameRootFields?: string[];
	// the schemaExtension field is used to extend the generated GraphQL schema with additional types and fields
	// this is useful for specifying type definitions for JSON objects
	schemaExtension?: string;
	replaceCustomScalarTypeFields?: ReplaceCustomScalarTypeFieldConfiguration[];
}

export interface GraphQLIntrospection extends GraphQLUpstream, GraphQLIntrospectionOptions {
	isFederation?: boolean;
}

export interface GraphQLFederationUpstream extends Omit<Omit<GraphQLUpstream, 'introspection'>, 'apiNamespace'> {
	name?: string;
	loadSchemaFromString?: GraphQLIntrospectionOptions['loadSchemaFromString'];
	introspection?: IntrospectionFetchOptions & GraphqlIntrospectionHeaders;
}

export interface GraphQLFederationIntrospection extends IntrospectionConfiguration {
	upstreams: GraphQLFederationUpstream[];
	apiNamespace?: string;
}

export interface ReplaceCustomScalarTypeFieldConfiguration {
	entityName: string;
	fieldName: string;
	inputTypeReplacement?: string;
	responseTypeReplacement: string;
}

export interface DatabaseIntrospection extends IntrospectionConfiguration {
	databaseURL: InputVariable;
	apiNamespace?: string;
	// the schemaExtension field is used to extend the generated GraphQL schema with additional types and fields
	// this is useful for specifying type definitions for JSON objects
	schemaExtension?: string;
	replaceCustomScalarTypeFields?: ReplaceCustomScalarTypeFieldConfiguration[];
}

export interface PrismaIntrospection extends IntrospectionConfiguration {
	prismaFilePath: string;
	apiNamespace?: string;
	// the schemaExtension field is used to extend the generated GraphQL schema with additional types and fields
	// this is useful for specifying type definitions for JSON objects
	schemaExtension?: string;
	replaceCustomScalarTypeFields?: ReplaceCustomScalarTypeFieldConfiguration[];
}

export interface IntrospectionFetchOptions {
	disableCache?: boolean;
	pollingIntervalSeconds?: number;
}

export interface IntrospectionConfiguration {
	// id is the unique identifier for the data source
	id?: string;
	/**
	 * Timeout for network requests originated by this data source, in seconds.
	 *
	 * @remarks
	 * See {@link NodeOptions| the NodeOptions type} for more details.
	 *
	 * @defaultValue Use the default timeout for this node.
	 */
	requestTimeoutSeconds?: number;
	introspection?: IntrospectionFetchOptions;
}

export interface HTTPUpstream extends IntrospectionConfiguration {
	apiNamespace?: string;
	headers?: (builder: IHeadersBuilder) => IHeadersBuilder;
	authentication?: HTTPUpstreamAuthentication;
	mTLS?: HTTPmTlsConfiguration;
	/** HTTP(S) proxy to use, overriding the default one. To disable a global proxy
	 * set its value to null.
	 *
	 * @defaultValue undefined, which uses the global proxy defined in configureWunderGraphApplication()
	 */
	httpProxyUrl?: InputVariable | null;
}

export type HTTPmTlsConfiguration = {
	/**
	 * 	Private-key or environment variable name that stores the key
	 */
	key: InputVariable;
	/**
	 * 	X.509 (TLS/HTTPS) or environment variable name that stores the certificate
	 */
	cert: InputVariable;
	/**
	 * InsecureSkipVerify controls whether a client verifies the server's certificate chain and host name
	 * If InsecureSkipVerify is true, crypto/tls accepts any certificate presented by the server and any host name in that certificate.
	 * In this mode, TLS is susceptible to machine-in-the-middle attacks unless custom verification is used.
	 * This should be used only for testing
	 */
	insecureSkipVerify: boolean;
};

export type HTTPUpstreamAuthentication = JWTAuthentication | JWTAuthenticationWithAccessTokenExchange;

export interface JWTAuthentication {
	kind: 'jwt';
	secret: InputVariable;
	signingMethod: JWTSigningMethod;
}

export interface JWTAuthenticationWithAccessTokenExchange {
	kind: 'jwt_with_access_token_exchange';
	secret: InputVariable;
	signingMethod: JWTSigningMethod;
	accessTokenExchangeEndpoint: InputVariable;
}

export type JWTSigningMethod = 'HS256';

export interface GraphqlIntrospectionHeaders {
	headers?: (builder: IGraphqlIntrospectionHeadersBuilder) => IGraphqlIntrospectionHeadersBuilder;
}

export interface GraphQLUpstream extends HTTPUpstream {
	url: InputVariable;
	baseUrl?: InputVariable;
	path?: InputVariable;
	subscriptionsURL?: InputVariable;
	subscriptionsUseSSE?: boolean;
	introspection?: IntrospectionFetchOptions & GraphqlIntrospectionHeaders;
}

export interface OpenAPIIntrospectionFile {
	kind: 'file';
	filePath: string;
}

export interface OpenAPIIntrospectionString {
	kind: 'string';
	openAPISpec: string;
}

export interface OpenAPIIntrospectionObject {
	kind: 'object';
	openAPIObject: {};
}

export type OpenAPIIntrospectionSource =
	| OpenAPIIntrospectionFile
	| OpenAPIIntrospectionString
	| OpenAPIIntrospectionObject;

export interface OpenAPIIntrospection extends HTTPUpstream {
	source: OpenAPIIntrospectionSource;
	// statusCodeUnions set to true will make all responses return a union type of all possible response objects,
	// mapped by status code
	// by default, only the status 200 response is mapped, which keeps the GraphQL API flat
	// by enabling statusCodeUnions, you have to unwrap the response union via fragments for each response
	statusCodeUnions?: boolean;
	baseURL?: InputVariable;
	// the schemaExtension field is used to extend the generated GraphQL schema with additional types and fields
	// this is useful for specifying type definitions for JSON objects
	schemaExtension?: string;
	replaceCustomScalarTypeFields?: ReplaceCustomScalarTypeFieldConfiguration[];
}

export interface StaticApiCustom {
	data: ConfigurationVariable;
}

export interface RESTApiCustom {
	Fetch: FetchConfiguration;
	Subscription: SubscriptionConfiguration;
	DefaultTypeName: string;
	StatusCodeTypeMappings: StatusCodeTypeMapping[];
}

export interface DatabaseApiCustom {
	prisma_schema: string;
	graphql_schema: string;
	databaseURL: ConfigurationVariable;
	jsonTypeFields: SingleTypeField[];
	jsonInputVariables: string[];
}

export interface SubscriptionConfiguration {
	Enabled: boolean;
	PollingIntervalMillis?: number;
	SkipPublishSameResponse?: boolean;
}

export interface GraphQLApiCustom {
	Federation: {
		Enabled: boolean;
		ServiceSDL: string;
	};
	Fetch: FetchConfiguration;
	Subscription: {
		Enabled: boolean;
		URL: ConfigurationVariable;
		UseSSE: boolean;
	};
	UpstreamSchema: string;
	HooksConfiguration: GraphQLDataSourceHooksConfiguration;
	CustomScalarTypeFields: SingleTypeField[];
}

export interface GraphQLServerConfiguration extends Omit<GraphQLIntrospection, 'loadSchemaFromString'> {
	schema: GraphQLSchema | Promise<GraphQLSchema>;
}

export const introspectGraphqlServer = async (introspection: GraphQLServerConfiguration) => {
	const { schema, ...rest } = introspection;
	const resolvedSchema = (await schema) as GraphQLSchema;

	return introspectGraphqlWithCache({
		...rest,
		internal: true,
		loadSchemaFromString: () => printSchema(buildClientSchema(introspectionFromSchema(resolvedSchema))),
	});
};

export const introspect = {
	graphql: (introspection: Omit<GraphQLIntrospection, 'isFederation'>) => {
		return introspectGraphqlWithCache(introspection);
	},
	postgresql: introspectPostgresql,
	mysql: introspectMySQL,
	planetscale: introspectPlanetScale,
	sqlite: introspectSQLite,
	sqlserver: introspectSQLServer,
	mongodb: introspectMongoDB,
	prisma: introspectPrisma,
	federation: introspectFederation,
	openApi: introspectOpenApi,
	openApiV2: introspectOpenApiV2,
	soap: introspectSoap,
};

export const buildUpstreamAuthentication = (upstream: HTTPUpstream): UpstreamAuthentication | undefined => {
	if (upstream.authentication === undefined) {
		return undefined;
	}
	return {
		kind: upstreamAuthenticationKind(upstream.authentication.kind),
		jwtConfig:
			upstream.authentication.kind === 'jwt'
				? {
						secret: mapInputVariable(upstream.authentication.secret),
						signingMethod: upstreamAuthenticationSigningMethod(upstream.authentication.signingMethod),
				  }
				: undefined,
		jwtWithAccessTokenExchangeConfig:
			upstream.authentication.kind === 'jwt_with_access_token_exchange'
				? {
						accessTokenExchangeEndpoint: mapInputVariable(upstream.authentication.accessTokenExchangeEndpoint),
						secret: mapInputVariable(upstream.authentication.secret),
						signingMethod: upstreamAuthenticationSigningMethod(upstream.authentication.signingMethod),
				  }
				: undefined,
	};
};

export const buildMTLSConfiguration = (upstream: HTTPUpstream): MTLSConfiguration | undefined => {
	if (upstream.mTLS === undefined) {
		return undefined;
	}
	return {
		key: mapInputVariable(upstream.mTLS?.key || ''),
		cert: mapInputVariable(upstream.mTLS?.cert || ''),
		insecureSkipVerify: upstream.mTLS?.insecureSkipVerify || false,
	};
};

const upstreamAuthenticationSigningMethod = (signingMethod: JWTSigningMethod): SigningMethod => {
	switch (signingMethod) {
		case 'HS256':
			return SigningMethod.SigningMethodHS256;
		default:
			throw new Error(`JWT signing method unsupported: ${signingMethod}`);
	}
};

const upstreamAuthenticationKind = (kind: HTTPUpstreamAuthentication['kind']): UpstreamAuthenticationKind => {
	switch (kind) {
		case 'jwt':
			return UpstreamAuthenticationKind.UpstreamAuthenticationJWT;
		case 'jwt_with_access_token_exchange':
			return UpstreamAuthenticationKind.UpstreamAuthenticationJWTWithAccessTokenExchange;
		default:
			throw new Error(`upstreamAuthenticationKind, unsupported kind: ${kind}`);
	}
};
