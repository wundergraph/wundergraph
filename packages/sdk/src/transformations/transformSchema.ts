import { replaceCustomScalars } from './replaceCustomScalars';
import { replaceCustomNumericScalars } from './replaceCustomNumericScalars';

export type { ReplaceCustomScalarsResult } from './replaceCustomScalars';

export type { ReplaceCustomNumericScalarsResult, ArgumentReplacement } from './replaceCustomNumericScalars';

const transformSchema = {
	replaceCustomScalars,
	replaceCustomNumericScalars,
};

export default transformSchema;
