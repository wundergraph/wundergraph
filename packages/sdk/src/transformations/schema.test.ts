import { GraphQLIntrospection } from '../definition';
import transformSchema from './schema';
import { assert } from 'chai';

test('replaceCustomScalars', async () => {
	const schema = `
		schema {
 			query: Query
 			mutation: Mutation
		}

		type Query {
			landpad(id: ID!): Landpad
		}
		
		type Mutation {
			createLandpad(landpad: LandpadInput!): Landpad
		}

		scalar JSON
		scalar hasura_geography

		type Landpad {
  		attempted_landings: String
  		details: String
  		full_name: String
  		id: ID
  		landing_type: String
  		location: JSON
  		status: String
  		successful_landings: String
  		wikipedia: String
  		geo: hasura_geography
		}
		
		type Pod {
		 	id: ID
		 	geo: hasura_geography
		} 	
		
		type Capsule {
		  id: ID
		  location: JSON
		}
		
		input LandpadInput {
			attempted_landings: String
			details: String
			full_name: String
			id: ID
			landing_type: String
			location: JSON
			status: String
			successful_landings: String
			wikipedia: String
			geo: hasura_geography
		}
	`;

	const introspection: GraphQLIntrospection = {
		url: '',
		schemaExtension: `
		type Location {
		  latitude: Float
		  longitude: Float
		  name: String
		  region: String
		}
		input LocationInput {
		  latitude: Float
		  longitude: Float
		  name: String
		  region: String
		}
		type Geography {
		  lat: Float
		  lng: Float
		}
		input GeographyInput {
		  lat: Float
		  lng: Float
		}
		`,
		replaceCustomScalarTypeFields: [
			{
				entityName: 'Landpad',
				fieldName: 'location',
				responseTypeReplacement: 'Location',
				inputTypeReplacement: 'LocationInput',
			},
			{
				entityName: 'Landpad',
				fieldName: 'geo',
				responseTypeReplacement: 'Geography',
				inputTypeReplacement: 'GeographyInput',
			},
			{
				entityName: 'Capsule',
				fieldName: 'location',
				responseTypeReplacement: 'Location',
				inputTypeReplacement: 'LocationInput',
			},
		],
	};

	const { schemaSDL, customScalarTypeFields } = transformSchema.replaceCustomScalars(schema, introspection);
	assert.equal(customScalarTypeFields.length, 3);
	assert.deepEqual(customScalarTypeFields, [
		{
			typeName: 'Landpad',
			fieldName: 'location',
		},
		{
			typeName: 'Landpad',
			fieldName: 'geo',
		},
		{
			typeName: 'Capsule',
			fieldName: 'location',
		},
	]);

	expect(schemaSDL).toMatchSnapshot();
});
