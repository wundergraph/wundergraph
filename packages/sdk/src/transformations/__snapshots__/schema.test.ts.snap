// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`replaceCustomScalars 1`] = `
"schema {
  query: Query
  mutation: Mutation
}

type Query {
  landpad(id: ID!): Landpad
}

type Mutation {
  createLandpad(landpad: LandpadInput!): Landpad
}

scalar JSON

scalar hasura_geography

type Landpad {
  attempted_landings: String
  details: String
  full_name: String
  id: ID
  landing_type: String
  location: Location
  status: String
  successful_landings: String
  wikipedia: String
  geo: Geography
}

type Pod {
  id: ID
  geo: hasura_geography
}

type Capsule {
  id: ID
  location: Location
}

input LandpadInput {
  attempted_landings: String
  details: String
  full_name: String
  id: ID
  landing_type: String
  location: LocationInput
  status: String
  successful_landings: String
  wikipedia: String
  geo: GeographyInput
}

type Location {
  latitude: Float
  longitude: Float
  name: String
  region: String
}

input LocationInput {
  latitude: Float
  longitude: Float
  name: String
  region: String
}

type Geography {
  lat: Float
  lng: Float
}

input GeographyInput {
  lat: Float
  lng: Float
}"
`;
